File: algorithms/3_hard/S/00761_special_binary_string/solution.cpp
class Solution {
public:
    static string makeLargestSpecial(const string& s) {
        const size_t len{ s.size() };
        if (len <= 2) {
            return s;
        }
        size_t left{ 0 };
        int balance{ 0 };
        vector<string> parts{};

        for (size_t right{ 0 }; right < len; ++right) {
            if (s[right] == '1') {
                balance += 1;
            } else {
                balance -= 1;
            }

            if (balance == 0) {
                string temp;
                temp.reserve(right - left + 1);
                temp.push_back('1');
                temp += makeLargestSpecial(s.substr(left + 1, right - left - 1));
                temp.push_back('0');
                parts.push_back(std::move(temp));
                left = right + 1;
            }
        }

        ranges::sort(parts, std::greater<>());

        string res;
        res.reserve(len);

        for (const auto& part : parts) {
            res += part;
        }

        return res;
    }
};

File: algorithms/3_hard/S/00761_special_binary_string/solution.py
class Solution:
    def makeLargestSpecial(self, s: str) -> str:
        n: int = len(s)
        if n <= 2:
            return s
        left: int = 0
        balance: int = 0
        parts: List[str] = []

        for right in range(n):
            if s[right] == '1':
                balance += 1
            else:
                balance -= 1

            if balance == 0:
                parts.append("1" + self.makeLargestSpecial(s[left + 1: right]) + "0")
                left = right + 1

        parts.sort(reverse=True)
        return "".join(parts)
